### PR TITLE
Cartesian vs Next example

### PR DESCRIPTION
**Pull request by havogt**
Add an example illustrating using gt4py.cartesian and gt4py.next computations next to each other using gt4py.next storages.

Refactor GTFieldInterface and cleanup GTDimsInterface for next.

### CODE REVIEW

**Comment by egparedes** [comment]
I would wait to merge this after storages are merged so both versions can use similar allocation code.

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:588]: add GTFieldInterface

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:577]: I assume this Protocol should be used if users define their own buffers that should be treated as fields. `__gt_domain__` is not enough, we need to specify how to get the buffer.
- Additionally: do want models of the concept to be also directly usable as field? or should this interface only be used to extract a buffer (that we could then wrap (possibly with host<->device copies) into a proper Field for embedded). Probably the latter makes sense, as the former is the `Field` interface.
- Note: forcing to add a member to a user object that returns an instance of a gt4py class, doesn't feel nice. Maybe we should at least allow `-> DomainLike` for `__gt_domain__`. Additionally, we would need `Dimension` to be any hashable object (similar to xarray) to be compatible in that sense.
- We should implement the same utility function for `__gt_domain__` that we have for `__gt_dims__`: not access the property directly but via the function, then we can provide default implementation for known types or additionally provide a single dispatch point for extending the list of default implementations.

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:603]: this default implementation can live in `GTFieldInterface` as it can be derived from `__gt_domain__`

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:602]: this make `__gt_dims__` again the same return type as for cartesian which probably makes sense, gtx should use the `__gt_domain__` property.

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:588]:
```suggestion
    core_defs.GTFieldInterface, Protocol[DimsT, core_defs.ScalarT]
```

File: src/gt4py/storage/cartesian/utils.py
Comment (havogt) [inline src/gt4py/storage/cartesian/utils.py:195]: export `__dlpack__` from the underlying field buffer.

File: src/gt4py/next/__init__.py
Comment (havogt) [reply src/gt4py/next/__init__.py:44]: On a second thought, I think we should keep the very explicit names, because people (at this stage at least) should know what they are using and being able to communicate to us. If we go for a default and switch it later, it might add unnecessary confusion.

File: src/gt4py/storage/cartesian/utils.py
Comment (havogt) [reply src/gt4py/storage/cartesian/utils.py:195]: Not doing it now as it is part of our Field/Array/Buffer conversation.

File: src/gt4py/next/common.py
Comment (havogt) [inline src/gt4py/next/common.py:583]:
```suggestion
    - A default implementation of the `__gt_dims__` interface from `gt4py.cartesian` is provided.
```

File: src/gt4py/next/iterator/embedded.py
Comment (egparedes) [inline src/gt4py/next/iterator/embedded.py:686]: Just a question: if there is now a `_get_domain(foo)`, shouldn't then `_get_axes(foo)` be implemented as an alias to   `_get_domain(foo).ndims`?

**Review by egparedes — APPROVED** [review]
It looks good to me. I only have a couple of questions.

**Pull request merged**
